src/NHibernate/Driver/FirebirdClientDriver.cs

```diff
@@ -99,15 +115,27 @@ private HashSet<string> GetCastCandidates(string statement)
 			return new HashSet<string>(candidates);
 		}
 
-		private void TypeCastParam(DbParameter param, DbCommand command)
+		private void TypeCastParam(DbParameter param, DbCommand command, SqlType sqlType)
 		{
-			var castType = GetFbTypeFromDbType(param.DbType);
-			command.CommandText = command.CommandText.ReplaceWholeWord(param.ParameterName, string.Format("cast({0} as {1})", param.ParameterName, castType));
+			var castType = GetFbTypeForParam(sqlType);
+			command.CommandText = command.CommandText.ReplaceWholeWord(
+				param.ParameterName,
+				$"cast({param.ParameterName} as {castType})");
 		}
 
-		private string GetFbTypeFromDbType(DbType dbType)
+		private string GetFbTypeForParam(SqlType sqlType)
 		{
-			return _fbDialect.GetCastTypeName(new SqlType(dbType));
+			switch (sqlType.DbType)
+			{
+				case DbType.Currency:
+					// Take the dialect binding for currency: decimal(18,4)
+					return _fbDialect.GetTypeName(sqlType);
+				case DbType.Decimal:
+					if (sqlType.PrecisionDefined)
+						return _fbDialect.GetTypeName(sqlType);
+					return _fbDialect.GetTypeName(SqlTypeFactory.GetSqlType(sqlType.DbType, _dbDecimalPrecision, _dbDecimalScale));
+			}
+			return _fbDialect.GetCastTypeName(sqlType);
```

Comment:
This `GetCastTypeName` is quite a terrible method, hopefully having few usages. It disregards the type `Length`/`Precision`/`Scale` and use instead `Column.DefaultLength`/`Column.DefaultPrecision`/`Column.DefaultScale`, which are `255`/`19`/`2`.
This causes all string types to be cast to a `255` length. Decimal type are trim down to `18` precision by dialect, but anyway, since the type resolution currently compared provided length with the max for the type, which is well below `255` for `decimal`, it always uses the default for `decimal` type, which for Firebird is `decimal(18,5)`. Otherwise we would have got a truncation at 2 digits after dot.
I will fill another Jira for that method.